frame/clearing-house/src/lib.rs

```diff
@@ -957,27 +1068,130 @@ pub mod pallet {
 		}
 	}
 
-	// ----------------------------------------------------------------------------------------------------
-	//                                           Helper Functions
-	// ----------------------------------------------------------------------------------------------------
+	// ---------------------------------------------------------------------------------------------
+	//                                    Helper Functions
+	// ---------------------------------------------------------------------------------------------
 
 	// Helper functions - core functionality
 	impl<T: Config> Pallet<T> {
+		fn summarize_account_state(
+			account_id: &T::AccountId,
+		) -> Result<AccountSummary<T>, DispatchError> {
+			let mut summary = AccountSummary::<T>::default();
+			for position in Self::get_positions(account_id) {
+				let market = Self::try_get_market(&position.market_id)?;
+				if let Some(direction) = position.direction() {
+					// should always succeed
+					let (position_notional, pnl) =
+						Self::abs_position_notional_and_pnl(&market, &position, direction)?;
+
+					let margin_requirement =
+						position_notional.try_mul(&market.margin_ratio_maintenance)?;
+					let info = PositionInfo::<T> {
+						direction,
+						margin_requirement_maintenance: margin_requirement,
+						base_asset_value: position_notional,
+						unrealized_pnl: pnl,
+						unrealized_funding: <Self as Instruments>::unrealized_funding(
+							&market, &position,
+						)?,
+					};
+
+					summary
+						.margin_requirement_maintenance
+						.try_add_mut(&info.margin_requirement_maintenance)?;
+					summary.base_asset_value.try_add_mut(&info.base_asset_value)?;
+					summary.unrealized_pnl.try_add_mut(&info.unrealized_pnl)?;
+					summary.unrealized_funding.try_add_mut(&info.unrealized_funding)?;
+					summary.positions_summary.push((market, position, info));
+				}
+			}
+
+			Ok(summary)
+		}
+
+		/// Fully liquidates the user's positions until his account is brought back above the MMR
+		///
+		/// ## Storage modifications
+		///
+		/// - Updates the [`markets`](Markets) of closed positions (according to changes in
+		///   [`Self::close_position_in_market`]).
+		/// - Removes closed [`positions`](Positions)
+		/// - Updates the user's account [`collateral`](Collateral)
+		///
+		/// ## Returns
+		///
+		/// The total liquidation fee
+		fn fully_liquidate_account(
+			user_id: &T::AccountId,
+			summary: AccountSummary<T>,
+		) -> Result<T::Balance, DispatchError> {
+			let mut collateral = Self::get_collateral(user_id).unwrap_or_else(Zero::zero);
+			let mut margin: T::Decimal = Self::updated_balance(
```

Comment:
Nitpick: `T::Decimal` is redundant here since you call `into_decimal in line 1134, right?

Reply:
Good catch! I thought `clippy` would warn me of redundant type annotations...

Reply:
Although `T::Decimal` is not always redundant... There are other instances that you call `into_decimal` but `T::Decimal` is still required by the compiler.

That could be the case for clippy not warning this?